core/src/main/java/org/apache/calcite/sql/validate/implicit/AbstractTypeCoercion.java

```diff
@@ -274,9 +274,10 @@ protected boolean needToCast(SqlValidatorScope scope, SqlNode node,
       return false;
     }
 
-    // No need to cast between char and varchar.
+    // No need to cast between char and unlimited varchar.
     if (SqlTypeUtil.isCharacter(toType)
-        && SqlTypeUtil.isCharacter(fromType)) {
+        && SqlTypeUtil.isCharacter(fromType)
+        && toType.getPrecision() == RelDataType.PRECISION_NOT_SPECIFIED) {
```

Comment:
Why do we need to judge precision?

Reply:
@caicancai `Character`includes char, varchar, longvarchar. Are these what you want?

Reply:
clearly, the original code is wrong. The comment says "varchar", but the code does not check for varchar. Here I have made the code match the comment, which seems correct. I haven't investigated what happens to binary objects. It's possible that this check is still not strong enough, but it is certainly better than it was, and it fixes the problem reported.

In general, when fixing a bug, I am trying to make the minimal changes necessary to solve the problem reported, without breaking anything else. Calcite is very complex, as you can see, even a small change may affect lots of tests.